Internalise the NomicBERT model

ed22699 opened the pull request:
# What does this PR do?

This PR internalises the NomicBERT model, following the basic structure of the https://huggingface.co/nomic-ai/nomic-bert-2048 

Fixes #42738

## Problem
BERT-like models using RoPE are currently not internalized in our codebase, e.g. https://huggingface.co/nomic-ai/nomic-bert-2048

## Solution
This PR creates a basic internalized version of nomic-bert-2048 with required modifications.

- [x] **Modular file**: `modular_nomic_bert.py` implemented and verified with `python utils/modular_model_converter.py modular_nomic_bert.py`
- [x] **Conversion script**: `convert_nomic_bert_to_hf.py` added with usage examples
- [x] **Integration tests**: End-to-end tests with exact output matching (text or logits)
- [x] **Documentation**: Model docs added/updated in `docs/source/en/model_doc/`
- [x] **Pattern reuse**: Verified against similar models (LLaVA, Idefics2, etc.)
- [x] **Quality checks**: `make fixup` passes with no errors

## Who Can Review?
@ArthurZucker @Cyrilvallez (text models)

ed22699 committed:
Created nomicBert skeleton

ed22699 committed:
Implement modular_nomic_bert

Co-authored-by: Felix Arkle <felixarkle@icloud.com>

ed22699 committed:
Implement convert_nomic_bert_to_hf

Bumsparkle committed:
Complete nomic_bert conversion script key mappings and fix NomicBertSelfAttention signature

ed22699 committed:
Create nomic bert documentation

Implemented descriptions for the main nomic bert documentation and
debugged modular_nomic_bert

ed22699 committed:
Remove redundancies and improve documentation for nomic_bert

Co-authored-by: Felix Arkle <felixarkle@icloud.com>

ed22699 committed:
Update dependencies for nomic_bert

Add einops to setup and add availibility checks for more graceful exit
if not available

ed22699 committed:
Fix nomic_bert attention mechanism

previous version overrote bert, leading to forward_unimplemented

ed22699 committed:
Correct past_key_value to past_key_values in NomicBertSelfAttention

ed22699 committed:
Implement cache_position into NomicBertSelfAttention

ed22699 committed:
Implement transpose_for_scores in NomicBertSelfAttention

ed22699 committed:
Fix nomicBertSelfAttention

Remove code which broke the encoder only assumption

ed22699 committed:
Add kwargs to NomicBertSelfAttention and ignore non-encoder logic

ed22699 committed:
Add past_key_values to NomicBertSelfAttention output

ed22699 committed:
Alter head dimension logic for NomicBertSelfAttention

Alter logic so smaller hidden dimensions are still computed correctly
and not lost

ed22699 committed:
Attempt to reflect hidden dim in NomicBertSelfAttention output shape

ed22699 committed:
Attempt to format output head shape for NomicBertSelfAttention

ed22699 committed:
Add is_decoder check to NomicBertSelfAttention

Although NomicBERT is encoder only model BertGeneration also requires it
to have decoder capabilities

ed22699 committed:
Update NomicBertSelfAttention to handle dynamic cache

ed22699 committed:
Update layer_idx to be a valid integer in NomicBertSelfAttention

ed22699 committed:
Alter output value size for NomicBertSelfAttention

ed22699 committed:
Improve seq_len_offset and past key robustness for NomicBertSelfAttention

ed22699 committed:
Implement left-padded batch inference for Nomic_Bert

ed22699 committed:
Explicitly add helper functions to NomicBert

ed22699 committed:
Fix dynamic cache issues within modular_nomic_bert

ed22699 committed:
Debug key errors of modular_nomic_bert

ed22699 committed:
Attempt to prevent typeError noneType in modular_nomic_bert

ed22699 committed:
Fix cache use in modular_nomic_bert

ed22699 committed:
Remove past_key_value tensor logic in modular_nomic_bert

Only use instances of Cache objects over tensor objects when dealing
with pask_key_value

ed22699 committed:
Convert legacy tensor format to Cache in modular_nomic_bert

ed22699 committed:
Implement reorder_cache for nomic_bert_model method

ed22699 committed:
Add none type safety to modular_nomic_bert

ed22699 committed:
Fix potential crop on none issues for modular_nomic_bert

ed22699 committed:
Fix RoPE offset misuse in modular_nomic_bert

ed22699 committed:
Use seqlen_offset for Cached decoding in NomicBertSelfAttention

ed22699 committed:
Handle any attention mask mismatches in NomicBertSelfAttention

ed22699 committed:
Update pretrain paths in test_modeling_nomic_bert

ed22699 committed:
Make test_modeling_nomic_bert more specific to nomic bert

ed22699 committed:
Prioritise position_ids over offset in NomicBertSelfAttention

ed22699 committed:
Remove noneType error from NomicBertEncoder

ed22699 committed:
Fix 1 is not greater than 1 error in modular_nomic_bert

ed22699 committed:
Update test suite to be nomic bert suitable

Co-authored-by: Felix Arkle <felixarkle@icloud.com>

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Update NomicBERT to comply with new hugging face standards

ed22699 committed:
Add dates to nomic_bert.md

ed22699 committed:
Resolve RoPE issues with large decoder inputs for NomicBERT

vasqu commented on docs/source/en/model_doc/nomic_bert.md:
```diff
@@ -0,0 +1,153 @@
+from transformers import pipeline
+
+pipeline = pipeline(
+    task="fill-mask",
```
cc @Rocketknight1 re #43325

Same usage as in normal Bert so I don't think we should deprecate

vasqu commented on docs/source/en/model_doc/nomic_bert.md:
```diff
@@ -0,0 +1,153 @@
+## NomicBertLMHeadModel
+
+[[autodoc]] NomicBertLMHeadModel
+    - forward
```
Let's remove these, not really valuable docs and not a model class that's actually used (top level)

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+        rotary_emb_scale_base (`float`, *optional*):
+            Scale base for the rotary embeddings.
+        rotary_emb_interleaved (`bool`, *optional*, defaults to `False`):
+            Whether to use interleaved rotary embeddings.
```
This is probably custom RoPE, we will need to refactor this to something similar as llama https://github.com/huggingface/transformers/blob/314f10929a2215b74c2ad6ecf7b2f380c9b7468a/src/transformers/models/llama/modeling_llama.py#L72

Only the theta should be relevant, we have `default_theta` as attribute if it's different to `10_000`

vasqu commented on docs/source/en/model_doc/nomic_bert.md:
```diff
@@ -0,0 +1,153 @@
+
+## Notes
+
+- Inputs should be padded on the right because RoPE handles relative positions differently than absolute embeddings.
```
I would still recommend right padding as we will simplify our code based on that (generate would handle a lot in regards to this which is not possible here)

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+        # For RoPE, we usually want strictly incremental IDs 0, 1, 2... for the valid tokens.
+        incremental_indices = (torch.cumsum(mask, dim=1).type_as(mask) + past_key_values_length - 1) * mask
+
+        return incremental_indices.long()
```
Let's simplify as we assume right padding, i.e. remove this

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+    via `NomicBertSelfAttention` and `position_ids` propagation. It handles the specific
+    initialization requirements of NomicBERT while maintaining compatibility with the
+    Transformers library build system.
+    """
```
Can be removed, the doc doesnt add much value

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+
+
+class NomicBertCrossAttention(BertCrossAttention):
+    pass
```
This feature is very outdated, let's remove support for this

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+    Overrides standard BERT components to incorporate:
+    Rotary Positional Embeddings (RoPE) in the Attention mechanism.
+    And SwiGLU activation in the Intermediate layer.
+    """
```
Noting last time, but these docstrings add little value

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+        intermediate_output = self.intermediate(attention_output)
+        layer_output = self.output(intermediate_output, attention_output)
+        outputs = (layer_output,) + outputs
+        return outputs
```
We really should be able to completely use Bert except for passing position embeddings

There are super outdated features and ways to do it like head mask, position ids, adding output attentions

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+            hidden_states=all_hidden_states,
+            attentions=all_self_attentions,
+            cross_attentions=all_cross_attentions,
+        )
```
Again please take a look at the original Bert model, remove the cross attention / decoder (causal) logic and prepare the position embeddings there (see llama). 

This uses outdated way to do things; also see the decorators that are used there

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+            hidden_states=encoder_outputs.hidden_states,
+            attentions=encoder_outputs.attentions,
+            cross_attentions=encoder_outputs.cross_attentions,
+        )
```
Reiterating to please sync how Bert looks currently + remove the unnecessary logic (cross, causal etc)

vasqu reviewed (commented):
An initial review:
- RoPE needs to be refactored
- We should avoid the cross attention + decoder logic (unused/rare feature)
- Follow how modern implementations look like, e.g. bert + llama are good sources for this
- Mind the decorators, they remove a lot of the manual things like return dict, output attns etc

Also sorry for the late review and dw about CI, it's known :hugs:

ed22699 committed:
Update nomic_bert.md and remove einops dependency from setup.py

ed22699 commented on docs/source/en/model_doc/nomic_bert.md:
```diff
@@ -0,0 +1,153 @@
+<!--Copyright 2025 the HuggingFace Team. All rights reserved.
```
No worries, thanks for the review and happy new year! I have updated the md file and will look through the conversion file and modular file over the next few days

ed22699 commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+    r"""
+    This is the configuration class to store the configuration of a [`NomicBertModel`]. It is used to instantiate an NomicBERT
+    model according to the specified arguments, defining the model architecture. Instantiating a configuration with the
+    defaults will yield a similar configuration to that of the [nomic-ai/nomic-bert-2048](https://huggingface.co/nomic-ai/nomic-bert-2048).
```
Should I also update the `nomic_bert.md` and `convert_nomic_bert_to_hf` files to use nomic-embed-text-v1.5 for examples instead?

ed22699 committed:
Upate RoPE for NomicBERT

- Update RoPE to be consistent with llama
- Add the BertTokenizer to NomicBERT
- Remove some redundant features within modular_nomic_bert.py

ed22699 commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+        return outputs
+
+
+class RotaryEmbedding(nn.Module):
```
Have now refactored for llama style rope

ed22699 commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+            )
+
+    def _round_to_multiple(self, value: int, multiple: int) -> int:
+        return ((value + multiple - 1) // multiple) * multiple
```
Removed

ed22699 commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,1201 @@
+
+        self.gradient_checkpointing = False
+        # Use NomicBertLayer by default if not specified
+        layer_class = layer_class or NomicBertLayer
```
Removed

vasqu commented:
@ed22699 can you ping me when it's ready for another review? will ignore smaller updates for now

ed22699 committed:
Adapt NomicBERT to be more consistent with bert

- Adapted `NomicBertSelfAttention`, `NomicBertAttention`,
`NomicBertIntermediate`, `NomicBertLayer`, `NomicBertEncoder`,
`NomicBertModel`

ed22699 committed:
Update NomicBertModel to create positionEmbeddings

ed22699 committed:
Make RoPE parameters for NomicBERT more robust

Fallback to defaults in init if RoPE parameters are None

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Resolved syntax error in calling NomicBertAttention

ed22699 committed:
Update NomicBertSelfAttention

Make NomicBertSelfAttention more consistent with BertSelfAttention,
taking RoPE elements to be consistent with LlamaAttention

ed22699 committed:
Update test_modeling_nomic_bert

Remove tests relating to decoder logic

ed22699 committed:
Resolve repo consistency errors within NomicBERT

ed22699 committed:
Update eager_attention_forward in NomicBERT

ed22699 committed:
Resolve attention mask errors in NomicBERT

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Update convert_nomic_bert_to_hf

Utilise conversion_mapping

ed22699 committed:
Merge branch 'bert-rope-model' of github.com:ed22699/transformers into bert-rope-model

ed22699 committed:
Reformat NomicBERT to meet repo standards

vasqu commented on docs/source/en/model_doc/nomic_bert.md:
```diff
@@ -0,0 +1,143 @@
+## NomicBertModel
+
+[[autodoc]] NomicBertModel
+    - forward
```
Nit: I would move this on top as everyone builds upon this

vasqu commented on src/transformers/conversion_mapping.py:
```diff
@@ -256,6 +257,151 @@ def _build_checkpoint_conversion_mapping():
         ],
+        "nomic_bert": [
+            # 1. Embeddings & Pooler (No 'bert.' prefix)
+            WeightRenaming("embeddings.word_embeddings.weight", "embeddings.word_embeddings.weight"),
```
Seeing this a few times, but we don't need to add things where nothing changes - only those that have a different name or similar

vasqu commented on src/transformers/models/nomic_bert/__init__.py:
```diff
@@ -0,0 +1,28 @@
+# Copyright 2025 the HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2026 the HuggingFace Team. All rights reserved.
```

vasqu commented on src/transformers/models/nomic_bert/convert_nomic_bert_to_hf.py:
```diff
@@ -0,0 +1,86 @@
+            type_vocab_size=base_config.type_vocab_size,
+            pad_vocab_size_multiple=base_config.pad_vocab_size_multiple,
+            tie_word_embeddings=base_config.tie_word_embeddings,
+            max_position_embeddings=base_config.max_position_embeddings,
```
Same here, these should be already defined no?

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,890 @@
+)
+
+
+class NomicBertConfig(BertConfig):
```
We also need to inherit from `RotaryEmbeddingConfigMixin` - might break modular so possibly needs to do everything manually :cry:

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,890 @@
+                "rope_theta": rotary_emb_base,
+            }
+        else:
+            self.rope_parameters = rope_parameters
```
The rotary mixin will do this for us already

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,890 @@
+    pass
+
+
+class NomicBertRotaryEmbedding(nn.Module):
```
```suggestion
class NomicBertRotaryEmbedding(LlamaRotaryEmbedding):
    pass
```
We should not need to manually do this, it's the standard rope as far as I can see

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,890 @@
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
+
+
+def eager_attention_forward(
```
No need to implement yourself again, you can import it from some other model and modular will unroll it in the modeling file, e.g. `from .. import eager_attention_forward`

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,890 @@
+        # Fallback
+        elif position_ids is not None and isinstance(position_ids, tuple):
+            cos, sin = position_ids
+            query_layer, key_layer = apply_rotary_pos_emb(query_layer, key_layer, cos, sin)
```
Should be similar to https://github.com/huggingface/transformers/blob/be87564082c7d0fc1c6508652b4490ecc89ad9ff/src/transformers/models/llama/modeling_llama.py#L260-L268
except for the fused projection then (position ids should not be used for cos, sin)

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,890 @@
+                if position_embeddings is not None:
+                    cos, sin = position_embeddings
+                    cache_kwargs = {"sin": sin, "cos": cos, "cache_position": cache_position}
+            key_layer, value_layer = past_key_values.update(key_layer, value_layer, self.layer_idx, cache_kwargs)
```
also cache related arguments should not be needed in the config along past_kv etc

iirc albert does not use a cache

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,890 @@
+                **kwargs,
+            )
+
+        attn_output = attn_output.transpose(1, 2).contiguous().view(batch_size, seq_len, -1)
```
https://github.com/huggingface/transformers/blob/be87564082c7d0fc1c6508652b4490ecc89ad9ff/src/transformers/models/llama/modeling_llama.py#L275-L292

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,890 @@
+            - **attentions** (`tuple(torch.FloatTensor)`, *optional*, returned when `output_attentions=True` is passed or when `config.output_attentions=True`) -- Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, num_heads, sequence_length, sequence_length)`.
+            - **cross_attentions** (`tuple(torch.FloatTensor)`, *optional*, returned when `output_attentions=True` and `config.add_cross_attention=True` is passed or when `config.output_attentions=True`) -- Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, num_heads, sequence_length, sequence_length)`.
+            - **past_key_values** (`tuple(tuple(torch.FloatTensor))`, *optional*, returned when `use_cache=True` is passed or when `config.use_cache=True`) -- Tuple of `torch.FloatTensor` of length `config.n_layers`, with each tuple containing the cached key and value states of the self-attention blocks.
+        """
```
Again autodocstring is your friend here (decorator)

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,890 @@
+                EncoderDecoderCache(DynamicCache(config=self.config), DynamicCache(config=self.config))
+                if encoder_hidden_states is not None or self.config.is_encoder_decoder
+                else DynamicCache(config=self.config)
+            )
```
No caching logic, albert should give a good basis

vasqu reviewed (commented):
Looks already much better! We can still go a bit deeper to make modular a bit better used + remove more unnecessary logic (caching, encoder_hidden_states (cross attn))

Also, no worries, take your time. We are not in a rush :hugs: have you checked that the model still produces similar outputs? It's important to keep close to the original model

ed22699 committed:
Implement recommendations into modular_nomic_bert

ed22699 committed:
Update NomicBert.md and convert_nomic_bert_to_hf

ed22699 commented:
Hi @vasqu made the suggested changes

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Fix issues found in modular_nomic_bert

ed22699 commented:
> Hi @vasqu made the suggested changes

Apologies misread tests_torch, still have 10 more tests to resolve before it is ready for review

ed22699 committed:
Update test_modeling_nomic_bert to use nomic-embed-text-v1.5

Was originally nomic-bert-2048

ed22699 committed:
Alter modular_nomic_bert to avoid legacy naming collisions

ed22699 committed:
Merge branch 'main' into bert-rope-model

vasqu commented on src/transformers/conversion_mapping.py:
```diff
@@ -316,6 +321,36 @@ def _build_checkpoint_conversion_mapping():
+                r"encoder.layers.(\d+).norm2",
+                r"encoder.layer.\1.output.LayerNorm",
+            ),
+            # QKV Splits (currently not passed)
```
```suggestion
```

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,654 @@
+
+        kwargs["is_decoder"] = kwargs.get("is_decoder", False)
+        kwargs["add_cross_attention"] = kwargs.get("add_cross_attention", False)
+        kwargs["use_cache"] = kwargs.get("use_cache", False)
```
If we use modular inheritance you can delete these with `del self.is_decoder` etc after the super init

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,654 @@
+            config.hidden_size,
+            self.num_kv_heads * self.attention_head_size,
+            bias=config.attention_bias,
+        )
```
Sorry but have you checked with the original model and if it produces similar results? The original has fused weights here

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,654 @@
+        return output
+
+
+class NomicBertIntermediate(nn.Module):
```
I would rather use https://github.com/huggingface/transformers/blob/b98123c4966c4821262a312c735c9946edab898b/src/transformers/models/llama/modeling_llama.py#L170
and rearrange a bit

We don't really need `NomicBertOutput` then and add the dropout etc to the layer

vasqu reviewed (commented):
Thanks for iterating, it's slowly taking shape. I added some comments which do stray a bit from the original bert but is a better standard to follow

ed22699 commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,654 @@
+class NomicBertPreTrainedModel(BertPreTrainedModel):
+    config_class = NomicBertConfig
+    base_model_prefix = "nomic_bert"
+    supports_gradient_checkpointing = False
```
looking at Bert it has its `base_model_prefix = "bert"`, when I break from that pattern and use `"model"` in my inherited version I seem to get the following test failure: 

NotImplementedError: `get_input_embeddings` not auto‑handled for NomicBertLMHeadModel; please override in the subclass.

ed22699 commented on docs/source/en/model_doc/nomic_bert.md:
```diff
@@ -0,0 +1,144 @@
+
+## Notes
+
+- For extremely long sequences, consider batching (gradient checkpointing is currently not supported) to save memory.
```
As NomicBertLayer inherits from `GradientCheckpointingLayer` I have now set `supports_gradient_checkpointing = True` and have added `self.gradient_checkpointing = False` into the Encoder init, similar to others in the repo. I am assuming this means I have added in the structure for users to support it themselves but have not explicitly implemented it to stay consistent with the repo. Is it best to just remove the bracketed section of the note and leave it there?

ed22699 committed:
Refactor NomicBERT to be more consistent with HF standards

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,582 @@
+
+        self.rotary_emb = NomicBertRotaryEmbedding(config)
+
+        self.post_init()
```
```suggestion
        self.rotary_emb = NomicBertRotaryEmbedding(config)
```
we only need to add the rope embeddings no?

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,582 @@
+
+        if position_ids is None:
+            position_ids = torch.arange(seq_length, dtype=torch.long, device=device)
+            position_ids = position_ids.unsqueeze(0).expand(batch_size, -1)
```
```suggestion
            position_ids = torch.arange(seq_length, dtype=torch.long, device=device)[None, :]
```
We don't need to expand to the batch size or do we?

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,582 @@
+            attention_mask=attention_mask,
+        )
+
+        position_embeddings = self.rotary_emb(embedding_output, position_ids)
```
```suggestion
        attention_mask = create_bidirectional_mask(
            config=self.config,
            input_embeds=embedding_output,
            attention_mask=attention_mask,
        )

        hidden_states = embedding_output
        position_embeddings = self.rotary_emb(hidden_states, position_ids)
```
just to be a bit more consistent with llama

vasqu commented on src/transformers/conversion_mapping.py:
```diff
@@ -316,6 +320,31 @@ def _build_checkpoint_conversion_mapping():
+                r"encoder.layers.(\d+).norm1",
+                r"encoder.layer.\1.attention.output.LayerNorm",
+            ),
+        ],
```
Are all weights loading as expected? Because looking at the original weights 

<img width="671" height="74" alt="Image" src="https://github.com/user-attachments/assets/d0aaed19-bc66-4738-815d-292bd4babac5" />

the modeling code does not match https://github.com/ed22699/transformers/blob/e9aa19e0ab042ab5ff01cab2f7a246909b9bfc7e/src/transformers/models/nomic_bert/modeling_nomic_bert.py#L270-L272

vasqu replied:
See my comments on the modules imo it's better to restructure a bit to have the out projection within the attention module and skip a few unnecessary intermediate modules

vasqu reviewed (commented):
There is still a bit room to improve the structure to follow Llama a bit more. Also better to merge/rebase since a few things might have changed in the meantime

Does the model load successfully? The attention weights don't seem to match the implementation

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Refactor Nomic Bert

ed22699 commented on docs/source/en/model_doc/nomic_bert.md:
```diff
@@ -0,0 +1,144 @@
+import torch
+from transformers import pipeline
+
+pipeline = pipeline(
```
Looking closer at the examples I realised they are still `nomic-bert-2048` examples so I have changed them to equal the 4 use cases listed in `nomic-embed-text-v1.5` instead

ed22699 commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,582 @@
+        attention_interface: Callable = eager_attention_forward
+        if self.config._attn_implementation != "eager":
+            attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
+
```
Moving to this new version produced 4 new failures:

================================================= short test summary info ==================================================
FAILED tests/models/nomic_bert/test_modeling_nomic_bert.py::NomicBertModelTest::test_eager_matches_sdpa_inference_10_fp32_pad_left_no_attn_mask_sdpa_kernels - ValueError: mean relative difference for hidden_states: 1.538e-02, torch atol = 1e-06, torch rtol = 0.0001
FAILED tests/models/nomic_bert/test_modeling_nomic_bert.py::NomicBertModelTest::test_eager_matches_sdpa_inference_11_fp32_pad_left_no_attn_mask - ValueError: mean relative difference for hidden_states: 1.539e-02, torch atol = 1e-06, torch rtol = 0.0001
FAILED tests/models/nomic_bert/test_modeling_nomic_bert.py::NomicBertModelTest::test_eager_matches_sdpa_inference_14_fp32_pad_right_no_attn_mask_sdpa_kernels - ValueError: mean relative difference for hidden_states: 1.915e-02, torch atol = 1e-06, torch rtol = 0.0001
FAILED tests/models/nomic_bert/test_modeling_nomic_bert.py::NomicBertModelTest::test_eager_matches_sdpa_inference_15_fp32_pad_right_no_attn_mask - ValueError: mean relative difference for hidden_states: 1.405e-02, torch atol = 1e-06, torch rtol = 0.0001
================================= 4 failed, 102 passed, 94 skipped, 25 warnings in 32.78s ==================================

will look at them tomorrow, apart from this everything else has been resolved

ed22699 committed:
Resolve NomicBert testEager value error

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Merge branch 'main' into bert-rope-model

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,503 @@
+        )
+
+        del self.is_decoder
+        del self.is_causal
```
Still need this overwritten, just as `self.is_causal=False` (we use this for e.g. flash attention to switch between causal or not)

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,503 @@
+
+        if attention_mask is None:
+            attention_mask = torch.ones((batch_size, seq_length), device=device)
+        attention_mask = self.get_extended_attention_mask(attention_mask, (batch_size, seq_length))
```
Yea I'd like to avoid using this and switch to the new mask API instead - mainly due to it being deprecated by me in another PR :D if the tols are too little we can override that test and set them a bit higher; suspecting the norms to create a bigger diff here

vasqu commented on tests/models/nomic_bert/test_modeling_nomic_bert.py:
```diff
@@ -0,0 +1,622 @@
+
+@require_torch
+class NomicBertModelTest(ModelTesterMixin, PipelineTesterMixin, unittest.TestCase):
+    all_generative_model_classes = ()
```
```suggestion
```
not 100% sure but I think we can remove since only `GenerationTesterMixin` would pick this up

vasqu commented on tests/models/nomic_bert/test_modeling_nomic_bert.py:
```diff
@@ -0,0 +1,622 @@
+            "feature-extraction": NomicBertModel,
+            "token-classification": NomicBertForTokenClassification,
+            "zero-shot": NomicBertForSequenceClassification,
+        }
```
Should be roughly the same as https://github.com/huggingface/transformers/blob/9f43e7b4e2e92044e56f4c68a60c17da1a514fed/tests/models/bert/test_modeling_bert.py#L449-L457
(minus the `text-generation`)

vasqu commented on tests/models/nomic_bert/test_modeling_nomic_bert.py:
```diff
@@ -0,0 +1,622 @@
+        model = NomicBertModel.from_pretrained(model_name)
+        self.assertIsNotNone(model)
+
+    def attention_mask_padding_matches_padding_free_with_position_ids(
```
Can be removed, I merged something to main that makes this compatible OOB

vasqu reviewed (commented):
Finally got to it :) left some comments, mainly on attention and convention styles

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Alter NomicBertSelfAttention to inherit from LlamaAttention

ed22699 commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,503 @@
+        )
+
+        # is_decoder not deleted as will cause errors if NomicBertForMaskedLM is called
+        self.is_decoder = False
```
It breaks when NomicBertForMaskedLM is called such as when you run the original nomic-bert-2048 example:
```python
config = AutoConfig.from_pretrained('nomic-ai/nomic-bert-2048', trust_remote_code=True) # the config needs to be passed in
model = AutoModelForMaskedLM.from_pretrained('nomic-ai/nomic-bert-2048',config=config, trust_remote_code=True)
```
my modular inherits from bert:
```python
class NomicBertForMaskedLM(BertForMaskedLM):
    pass
```
and the modellings init created becomes this: 
```python
@auto_docstring
class NomicBertForMaskedLM(NomicBertPreTrainedModel):
    _tied_weights_keys = {
        "cls.predictions.decoder.weight": "nomic_bert.embeddings.word_embeddings.weight",
        "cls.predictions.decoder.bias": "cls.predictions.bias",
    }

    def __init__(self, config):
        super().__init__(config)

        if config.is_decoder:
            logger.warning(
                "If you want to use `NomicBertForMaskedLM` make sure `config.is_decoder=False` for "
                "bi-directional self-attention."
            )

        self.nomic_bert = NomicBertModel(config, add_pooling_layer=False)
        self.cls = NomicBertOnlyMLMHead(config)

        # Initialize weights and apply final processing
        self.post_init()
```
however with nomic-text-embed-v1.5 preconfig this does not work anyway as it does not have any CLS weights

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Update docs date for nomic_bert.md

ed22699 committed:
Merge branch 'main' into bert-rope-model

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,480 @@
+        self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
+
+
+class NomicBertLayer(GradientCheckpointingLayer):
```
We can inherit from the gpt neox layer, at least for the init it seems

vasqu reviewed (commented):
Finally came back around to check, I have some smaller comments but overall I feel like we are pretty close. 

Could it be that some embedding dropout and layernorm is missing?

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,480 @@
+
+@auto_docstring
+class NomicBertModel(BertModel):
+    def __init__(self, config, add_pooling_layer=True):
```
Woops yes, I have changed this

vasqu commented on docs/source/en/model_doc/nomic_bert.md:
```diff
@@ -0,0 +1,198 @@
+    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)
+
+tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+model = AutoModel.from_pretrained("nomic-ai/nomic-embed-text-v1.5", trust_remote_code=True)
```
Imo, we can alreay open a PR and make the adjusted changes already. You can add `revision=f"refs/pr/{your_pr_number}"`

This will make it easy for us to ensure we have everything working

ed22699 committed:
Add missing embedding dropout to Nomic Bert

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 commented:
@vasqu think we're ready for another review, theres quite a lot of failures but they all seem to be coming from other models

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,550 @@
+        self.o_proj = nn.Linear(config.num_attention_heads * self.head_dim, config.hidden_size, bias=False)
+
+        self.is_causal = False
+        del self.is_decoder
```
```suggestion
    def __init__(self, config):
        super().__init__(config, layer_idx=None)
        del self.layer_idx

        self.attention_dropout = config.attention_probs_dropout_prob
        self.is_causal = False
```
This way, we have an easier time if we use the mistral one

vasqu commented on tests/models/nomic_bert/test_modeling_nomic_bert.py:
```diff
@@ -0,0 +1,404 @@
+
+        text = "Plants create oxygen through a process known as photosynthesis."
+        inputs = tokenizer(text, return_tensors="pt", padding=True, truncation=True)
+        inputs["attention_mask"] = torch.ones_like(inputs["input_ids"])
```
Let's make this a batch, e.g. two texts, and have a proper mask so we can check against padding working correctly

vasqu reviewed (commented):
Getting closer! There are a few things we could improve with the usage of modular but overall the core is solid, thanks a lot for iterating

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Reformat Nomic Bert config to be consistent with new standards

vasqu reviewed (approved):
Careful approval BUT
- We need a hub PR to at last properly verify and check against
- Maybe wait on jina v3 to simplify parts of the modular code

Awesome work, glad to have this come around 🫡

vasqu commented:
Btw main is stable again, you can merge/rebase :hugs:

ed22699 committed:
Merge branch 'main' into bert-rope-model

ed22699 committed:
Refactor nomic bert to inherit more from Jina V3

vasqu committed:
fixup jina v3 (doesnt use token type ids it seems) and inherit more fixups for nomic bert

vasqu commented:
@ed22699 I've pushed some small changes in 5f8d1c6bac8ebe3dec265ad8ba5ce3ff4f51f91f

Just some more modular stuff (small tricks ;)) - let me know when you are ready with the hub PR, will check if I can reach the authors then (if everything works re integration tests)

vasqu commented on src/transformers/models/nomic_bert/modular_nomic_bert.py:
```diff
@@ -0,0 +1,385 @@
+        if self.head_dim is None:
+            self.head_dim = self.hidden_size // self.num_attention_heads
+        if self.num_key_value_heads is None:
+            self.num_key_value_heads = self.num_attention_heads
```
Ah yea check out my latest commit, that shows how you can do it --> tl;dr: Use pretrained config for the init

ed22699 committed:
Update nomic_bert.md

Update examples to no longer cross reference bert in their tokenizer

vasqu commented on tests/models/nomic_bert/test_modeling_nomic_bert.py:
```diff
@@ -0,0 +1,372 @@
+    @slow
+    def test_inference_no_head_absolute_embedding(self):
+        model = AutoModel.from_pretrained("nomic-ai/nomic-embed-text-v1.5")
+        tokenizer = AutoTokenizer.from_pretrained("nomic-ai/nomic-embed-text-v1.5")
```
No revision, was it tested against the remote code?

vasqu reviewed (approved):
Sorry was off for a few days, careful approval, just a few smaller things - can you check?

vasqu committed:
quick fixups

vasqu committed:
Merge branch 'main' into bert-rope-model

vasqu committed:
fix

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_43067). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

vasqu committed:
fixes to align with original model

vasqu committed:
fixups

vasqu committed:
fix

vasqu committed:
remove conversion script, revert changes in conversion mapping

vasqu committed:
a10 numbers with remote model

vasqu committed:
fixup all other classes and conversion

vasqu committed:
fix

vasqu commented:
Gotcha, no worries. Good luck on your dissertation! :saluting_face:

vasqu committed:
new hub revision

vasqu committed:
Merge branch 'main' into bert-rope-model

vasqu commented:
run-slow: jina_embeddings_v3, nomic_bert

tomaarsen reviewed (commented):
My understanding is that this incorporates only the non-MoE path? The https://huggingface.co/nomic-ai/nomic-bert-2048/blob/main/modeling_hf_nomic_bert.py modeling code is used for various models, including:
* https://huggingface.co/nomic-ai/nomic-embed-text-v1
* https://huggingface.co/nomic-ai/nomic-embed-text-v1.5
* https://huggingface.co/nomic-ai/nomic-embed-text-v2-moe (but it uses MoE parameters, 8 experts, etc.)

These vision models:
* https://huggingface.co/nomic-ai/nomic-embed-vision-v1.5
* https://huggingface.co/nomic-ai/nomic-embed-vision-v1

And these research checkpoints:
* https://huggingface.co/nomic-ai/nomic-embed-text-v1-unsupervised
* https://huggingface.co/nomic-ai/nomic-embed-text-v1-ablated
* https://huggingface.co/nomic-ai/nomic-embed-text-v2-moe-unsupervised
* https://huggingface.co/nomic-ai/nomic-xlm-2048

I assume that this work is only aiming for the text portion. That does mean that we're diverging from the original implementation a bit, which also supports vision and MoE. Not strictly an issue, just something to note.
If we move forward, let's try to support not just the v1.5, but also the v1, it's also getting used a lot.

vasqu committed:
v1 tests, same code - slightly different config (on the hub)

vasqu committed:
fix wrong defaults

vasqu committed:
numbers didnt change on a10

vasqu committed:
fix warning

vasqu committed:
update docs

tomaarsen reviewed (commented):
Small nits, the general gist is solid I think.

vasqu committed:
update docs per toms review
